src/c_ext/hamt/node.hpp

```diff
@@ -0,0 +1,518 @@
+#ifndef LEAF_HPP
+#define LEAF_HPP
+
+#include <string>
+#include <vector>
+#include <functional>
+
+#include "utils.hpp"
+
+typedef unsigned int u_int;
+
+using namespace std;
+
+enum types {
+    EMPTY,
+    LEAF,
+    COLLISION,
+    INDEX,
+    INTERNAL
+};
+
+namespace hamt {
+    template <typename T>
+    class internal_node;
+
+    template <typename T>
+    class internal_node;
+
+    template <typename T>
+    class node {
+    protected:
+        types type;
+        u_int hash;
+        u_int mask;
+        u_int size;
+        string key;
+        T val;
+        bool can_edit;
+
+        vector<node<T> * > *chdrn;
+
+    public:
+        node()
+        {
+            this->type = EMPTY;
+            this->hash = 0;
+            this->key = "";
+            this->mask = 0;
+            this->size = 0;
+            this->val = T();
+            this->make_mut(false);
+
+            this->chdrn = new vector<node<T> * >();
+        }
+
+        node(const node<T>& rhs)
+        {
+            if (this != &rhs) {
+                this->type = rhs.get_type();
+                this->hash = rhs.get_hash();
+                this->key = rhs.get_key();
+                this->mask = rhs.get_mask();
+                this->size = rhs.get_size();
+                this->val = rhs.get_val();
+                this->make_mut(rhs.editable());
+
+                this->chdrn = rhs.get_chdrn();
```

Comment:
But then it's the children's pointer that gets copied. What if `rhs` is destroyed?
`this->chdrn` will get invalidated.

Also, will you insert some vowels in there dammit.